servus/uri.cpp

```diff
@@ -0,0 +1,383 @@
+
+    if( pos == std::string::npos )
+    {
+        if( needsSeparator[part] )
+        {
+            output = "";
+            return true;
+        }
+    }
+    else if ( disallowed[part] &&
+              input.find_first_of( disallowed[part] ) < pos )
+    {
+        output = "";
+        return true;
+    }
+
+    // If the separator is not the first character, assert that parts requiring
+    // an initial character find it.
+    assert( !requireFirst[part] || pos == 0 || input[0] == requireFirst[part] );
+
+    // If the separator was found at pos == 0, the returned string is empty
+    assert( input.size() >= skip[part] );
+    output = input.substr( skip[part], pos - skip[part] );
+    input = pos == std::string::npos ?
+        "" : input.substr( pos + postSkip[part] );
+    return true;
+}
+
+void _parseAuthority( URIData& data, const std::string& auth )
+{
+    const size_t atPos = auth.find_first_of( '@' );
+    if( atPos != std::string::npos )
+        data.userinfo = auth.substr( 0, atPos );
+    const size_t hostPos =
+        atPos == std::string::npos ? 0 : atPos + 1;
+    const size_t colonPos =
+        auth.find_first_of( ':', hostPos );
+    if( colonPos != std::string::npos )
+        data.port = std::stoi( auth.substr( colonPos + 1 ));
+    // Works regardless of colonPos == npos
+    data.host = auth.substr( hostPos, colonPos - hostPos );
+    if( data.host.empty( ))
+        throw std::invalid_argument("");
+}
+
+void _parseQueryMap( URIData& data )
+{
+    // parse query data into key-value pairs
+    std::string query = data.query;
+    while( !query.empty( ))
+    {
+        const size_t nextPair = query.find( ',' );
+        if( nextPair == 0 )
+        {
+            query = query.substr( 1 );
+            continue;
+        }
+
+        const std::string pair = query.substr( 0, nextPair );
+        if( nextPair == std::string::npos )
+            query.clear();
+        else
+            query = query.substr( nextPair + 1 );
+
+        const size_t eq = pair.find( '=' );
+        if( eq == std::string::npos || eq == 0 )
+            continue;
+        data.queryMap[ pair.substr( 0, eq ) ] = pair.substr( eq + 1 );
+    }
+}
+
+void _tolower( std::string& str )
+{
+    std::transform( str.begin(), str.end(), str.begin(), ::tolower );
+}
+
+class URI
+{
+public:
+    explicit URI( const std::string& uri )
+    {
+        if( uri.empty( ))
+            return;
+
+        try
+        {
+            _parseURI( uri );
+        }
+        catch( ... )
+        {
+            throw uri_parse( uri );
+        }
+    }
+
+    URIData& getData() { return _uriData; }
+    const URIData& getData() const { return _uriData; }
+
+private:
+    URIData _uriData;
+
+    void _parseURI( const std::string& uri )
+    {
+        std::string input = uri;
+        URIPart part = SCHEME;
+        while( !input.empty( ))
+        {
+            switch(part)
+            {
+            case SCHEME:
```

Comment:
I remember we had a discussin whether to parse it or not, was not there the standard regex from GNU and MSVC2013 already does that ?